src/query_plan/operation.rs

```diff
@@ -473,6 +479,106 @@ fn is_deferred_selection(directives: &DirectiveList) -> bool {
+///
+/// __typename information can always be provided by any subgraph declaring that type. While this data can be
+/// theoretically fetched from multiple sources, in practice it doesn't really matter which subgraph we use
+/// for the __typename and we should just get it from the same source as the one that was used to resolve
+/// other fields.
+///
+/// In most cases, selecting __typename won't be a problem as query planning algorithm ignores "obviously"
+/// inefficient paths. Typically, querying the __typename of an entity is generally ok because when looking at
+/// a path, the query planning algorithm always favor getting a field "locally" if it can (which it always can
+/// for __typename) and ignore alternative that would jump subgraphs.
+///
+/// When querying a __typename after a @shareable field, query planning algorithm would consider getting the
+/// __typename from EACH version of the @shareable field. This unnecessarily explodes the number of possible
+/// query plans with some useless options and results in degraded performance. Since the number of possible
+/// plans doubles for every field for which there is a choice, eliminating unnecessary choices improves query
+/// planning performance.
+///
+/// It is unclear how to do this cleanly with the current planning algorithm, so this method is a workaround
+/// so we can efficiently generate query plans. In order to prevent the query planner from spending time
+/// exploring those useless __typename options, we "remove" the unnecessary __typename selections from the
+/// operation. Since we need to ensure that the __typename field will still need to be queried, we "tag"
+/// one of the "sibling" selections (using "attachement") to remember that __typename needs to be added
+/// back eventually. The core query planning algorithm will ignore that tag, and because __typename has been
+/// otherwise removed, we'll save any related work. As we build the final query plan, we'll check back for
+/// those "tags" and add back the __typename selections. As this only happen after the query planning
+/// algorithm has computed all choices, we achieve our goal of not considering useless choices due to
+/// __typename. Do note that if __typename is the "only" selection of some selection set, then we leave it
+/// untouched, and let the query planning algorithm treat it as any other field. We have no other choice in
+/// that case, and that's actually what we want.
+fn optimize_sibling_typenames(
+    normalized_selection_set: &mut NormalizedSelectionSet,
+    interface_types_with_interface_objects: &IndexSet<InterfaceTypeDefinitionPosition>,
+) -> Result<(), FederationError> {
+    let selections = &mut normalized_selection_set.selections;
+
+    let is_interface_object = interface_types_with_interface_objects
+        .iter()
+        .any(|intf| intf.type_name == normalized_selection_set.ty);
```

Comment:
Since this is a set, we could do something like this to make it O(1). Very minor knowing it's likely not going to be the most expensive thing in the world anyway.
```rust
let is_interface_object = interface_types_with_interface_objects.contains(&InterfaceTypeDefinitionPosition {
    type_name: normalized_selection_set.ty.clone(),
});
```